### ISSUE TITLE
Improve Documentation

### ISSUE THREAD

**Issue by maydayv7** [maydayv7/dotfiles#7]
Documentation of features is sparse and requires improvement. Also need to consider blogs

**Comment by maydayv7** [comment]
Website: https://github.com/maydayv7/website
Hosted: https://maydayv7.github.io

**Comment by mannp** [comment]
I second this, as I am trying to follow converting my existing nix config to your structure, but it is nearly impossible with so many things combined into one setup eg your website, radical zfs system erase etc.... can I even use this config on a standard luks config?

I have followed items 2 to 6, but I am unable to build my new config, as my user secret is missing.

I am unclear how to run, or even if your 'nixos' scripts are available on a host system, prior to swapping to your config or not?

I am not entirely clear of your intention to whether you want others to utilise your config structures, or you are just putting this out there to store your configs... :)

**Comment by maydayv7** [comment]
> I second this, as I am trying to follow converting my existing nix config to your structure, but it is nearly impossible with so many things combined into one setup eg your website, radical zfs system erase etc.... can I even use this config on a standard luks config?

The system erase is done by reverting the `/` partition to a blank snapshot on every boot, and necessary files are persisted across reboots using the [`impermanence`](https://github.com/nix-community/impermanence) module. It is a tad difficult to set it all up in one go and not recommended if you are just starting out (since each and every file/directory to persist must be identified and declared)

Haven't tried it with LUKS Encryption (since I don't really need it). Important thing to note is that in order to keep the config as device-agnostic as possible, the filesystems are created and mounted using partition labels instead of device-specific UIDs. You can check out [`partitions.sh`](https://github.com/maydayv7/dotfiles/blob/stable/scripts/partitions.sh) (The script that makes and formats all partitions on install) and modify [`filesystem.nix`](https://github.com/maydayv7/dotfiles/blob/stable/modules/hardware/filesystem.nix) according to your requirements

The reason why everything is being done in a huge mono-repo structure is that it allows for easier maintenance, reduces code duplication and allows me to keep track of everything in one place

> I have followed items 2 to 6, but I am unable to build my new config, as my user secret is missing.

Hmm, not sure what problem you're facing. You should a create a GPG Key for your user, include that GPG key in `.sops.yaml`, and create all the secrets as desired. [`sops-nix`](https://github.com/Mic92/sops-nix) is used for this secrets management so you can best refer to their documentation. Also, if you wish to use SSH Keys instead (some people consider `gnupg` unsafe/outdated), you could replicate such my setup using [`agenix`](https://github.com/ryantm/agenix), as I have done in c9351ddfb58be74436d6564ee4ef62733a5852e7 (but found it convenient to just use GPG keys I already had, so used `sops-nix` instead)

> I am unclear how to run, or even if your 'nixos' scripts are available on a host system, prior to swapping to your config or not?

You can run it using `nix run github:maydayv7/dotfiles#nixos -- help`. It isn't available on host system without switching to the configuration, and I've also not really tested if it all commands work properly on any system apart from mine (especially the `install`, `setup` and `iso` commands)

> I am not entirely clear of your intention to whether you want others to utilise your config structures, or you are just putting this out there to store your configs... :)

Well, this is mainly for my personal use, but I would be glad if it could help others *build their own* config based off of mine :)

The documentation efforts take a lot of time, and I'm currently busy with school-work

**Comment by mannp** [comment]
Thanks for the helpful response.

I am not installing from scratch, so I envisaged partitions.sh would not be needed/utilized, and I would update filesystem.nix to match my existing setup. 

Is the impermanence module 'always on'/ or can be disabled, as I am trying to just get a basic config going, without enabling all the bells and whistles you have going on.

The sops-nix config is where I am falling foul, as I already have gpg keys for all users and frankly, sops-nix, has so many ways to install, it mixes up age and gpg configs and I guess I need to understand what that is doing before going any further.

I'd prefer to use it, though for gpg, as I use ssh user certs for access to all signed servers, rather than individual ssh key pairs.

So the nixos script doesn't work without a completed local build.... I was trying to utlise 'nixos secret' as I was seemingly missing a secret for my new user.

Anyway, will continue with my existing configs and over time might look again at this.

Thanks for sharing :)

**Comment by maydayv7** [comment]
> Is the impermanence module 'always on'/ or can be disabled, as I am trying to just get a basic config going, without enabling all the bells and whistles you have going on.

Yeah, basically it's always on, but to disable it you would not to comment out the persistence configuration (`config.environment.persistence.<partition>`, `config.environment.persist` and `config.user.persist`)

> The sops-nix config is where I am falling foul, as I already have gpg keys for all users and frankly, sops-nix, has so many ways to install, it mixes up age and gpg configs and I guess I need to understand what that is doing before going any further.

Hmm yeah even I had faced problems while trying to initially set up `sops-nix`. The way I did was just to create/use a GPG Key, find it's signature using `gpg --list-keys`, creating `.sops.yaml` in `secrets`, appending my key to the file as specified in the `sops-nix` documentation, then pointing `sops` to that file (by prefixing every command using <code>--config <i>/path/to/<b>.sops.yaml</b></i></code>), and then creating the required secrets. An important thing to note is that `sops` requires the GPG Keys to be imported by the user who creates/edits/views the secrets but the `sops-nix` module requires that the GPG Keys are present on disk in a specified location (using `config.sops.gnupg.home`). For this, I have placed `pubring.kbx` and `private-keys-v1.d` in `files/gpg` from `$HOME/.gnupg` and pointed `sops.gnupg.home` to `/etc/nixos/files/gpg`, and encrypted those files with `git-crypt`, so that the `sops-nix` module finds it in the same location every boot and the secrets are decrypted properly. You do not need to set up `age` or SSH in any way if you want to just use it with GPG (although you do need to make sure that `config.sshKeyPaths` is set to `mkForce []`)

> So the nixos script doesn't work without a completed local build.... I was trying to utlise 'nixos secret' as I was seemingly missing a secret for my new user

It does work (but it assumes that the config lies in `/etc/nixos`), and also you would need to set `sops.secrets.<name>.neededForUsers = true` for the particular secret in case you are using it as `passwordFile` for the user
